Consolidate defaulting for ControllerManagerConfigurationSpec

tenzen-y opened the pull request:
Signed-off-by: tenzen-y <yuki.iwai.tz@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
As discussed in [this](https://github.com/kubernetes-sigs/kueue/pull/362#issuecomment-1232041979), I implemented the defaulting webhooks to set the default values for `ControllerManagerConfigurationSpec`.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

tenzen-y commented on main.go:
```diff
@@ -220,40 +220,31 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
 	}
 
-	config := config.Configuration{}
+	configuration := config.Configuration{}
```
Since `config` conflicts with `config "sigs.k8s.io/kueue/apis/config/v1alpha2`, I replaced `config` with `configuration`.

tenzen-y commented:
/assign @kerthcet

kerthcet commented:
I'll review this later.

kerthcet commented:
Sorry for let this sit for such a long time, just back from holiday. @tenzen-y

kerthcet commented on main.go:
```diff
@@ -220,40 +220,31 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
+			Port:                   config.DefaultWebhookPort,
+			HealthProbeBindAddress: config.DefaultHealthProbeBindAddress,
+			MetricsBindAddress:     config.DefaultMetricsBindAddress,
+			LeaderElectionID:       config.DefaultLeaderElectionID,
```
Just created a PR here https://github.com/kubernetes-sigs/kueue/pull/413, if leaderElection is false, leaderElectionID should be empty since we have no config file here.

tenzen-y replied:
It makes sense.
I removed this leaderElectionID.

tenzen-y commented on main.go:
```diff
@@ -220,40 +220,31 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
-		scheme.Default(&config)
+		scheme.Default(&configuration)
+		options = ctrl.Options{
+			Port:                   config.DefaultWebhookPort,
```
Thanks!

tenzen-y commented on apis/config/v1alpha2/defaults.go:
```diff
@@ -36,9 +41,31 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	if len(cfg.Health.HealthProbeBindAddress) == 0 {
+		cfg.Health.HealthProbeBindAddress = DefaultHealthProbeBindAddress
+	}
+	if cfg.LeaderElection == nil {
```
looks good!

tenzen-y commented:
@kerthcet 
Thanks for the review!
I addressed your review comments.

alculquicondor commented:
/retitle Consolidate defaulting for ControllerManagerConfigurationSpec

Note that there is no webhook involved in this PR, just API defaulting.

alculquicondor reviewed (commented):
Just nits

tenzen-y commented on apis/config/v1alpha2/defaults.go:
```diff
@@ -36,9 +40,28 @@ func addDefaultingFuncs(scheme *runtime.Scheme) error {
 
 // SetDefaults_Configuration sets default values for ComponentConfig.
 func SetDefaults_Configuration(cfg *Configuration) {
+
```
It makes sense.

tenzen-y replied:
Done.

kerthcet commented:
@tenzen-y can you rebase and address the comments, totally looks good to me.

tenzen-y commented on main.go:
```diff
@@ -85,7 +85,7 @@ func main() {
 	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
 
-	options, config := apply(configFile)
+	options, configuration := apply(configFile)
```
Done.

tenzen-y commented:
> @tenzen-y can you rebase and address the comments, totally looks good to me.

@kerthcet @alculquicondor I have rebased this PR and addressed comments by @alculquicondor.
PTAL

tenzen-y commented:
@kerthcet I found a forgotten correction in the below codes and fixed that. Could you please re-check this PR?

- Incorrect

https://github.com/kubernetes-sigs/kueue/blob/1b94fa5d8621b5f008cc66f67044642bd54f758f/main_test.go#L272-L274

- Correct

https://github.com/kubernetes-sigs/kueue/blob/1e1dfc11ac5ea057627f9645923a90f0b0d83221/main_test.go#L272-L274

alculquicondor commented on apis/config/v1alpha2/defaults_test.go:
```diff
@@ -36,8 +56,117 @@ func TestSetDefaults_Configuration(t *testing.T) {
+				},
+			},
+		},
+		"should not defaulting ControllerManagerConfigurationSpec": {
```
```suggestion
		"should not default ControllerManagerConfigurationSpec": {
```

tenzen-y replied:
Done.

alculquicondor commented on main.go:
```diff
@@ -220,40 +220,31 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
 	if configFile == "" {
-		scheme.Default(&config)
+		scheme.Default(&cfg)
+		options = ctrl.Options{
+			Scheme:                 scheme,
+			Port:                   config.DefaultWebhookPort,
+			HealthProbeBindAddress: config.DefaultHealthProbeBindAddress,
+			MetricsBindAddress:     config.DefaultMetricsBindAddress,
+			LeaderElection:         false,
+		}
```
I wonder if this works
```suggestion
		options, err = options.AndFrom(cfg)
```

tenzen-y replied:
That works fine on my local. Thanks!

alculquicondor commented on main.go:
```diff
@@ -220,40 +220,31 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
 			setupLog.Error(err, "unable to encode config file")
 			os.Exit(1)
 		}
 		setupLog.Info("Successfully loaded config file", "config", cfgStr)
```
We should also log when the configuration is just the default. It's useful for debugging.

tenzen-y replied:
Makes sense.

tenzen-y replied:
Done.

tenzen-y commented on main.go:
```diff
@@ -220,40 +220,36 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
+			setupLog.Error(err, "unable to encode the default config")
+			os.Exit(1)
+		}
+		setupLog.Info("Successfully loaded the default config", "config", cfgStr)
```
This logger output the below:

```shell
{"level":"info","ts":"2022-10-12T16:30:30.00241243Z","logger":"setup","caller":"logr@v1.2.3/logr.go:261","msg":"Successfully loaded the default config","config":"apiVersion: config.kueue.x-k8s.io/v1alpha2\nhealth:\n  healthProbeBindAddress: :8081\ninternalCertManagement:\n  enable: true\n  webhookSecretName: kueue-webhook-server-cert\n  webhookServiceName: kueue-webhook-service\nkind: Configuration\nmanageJobsWithoutQueueName: false\nmetrics:\n  bindAddress: :8080\nnamespace: kueue-system\nwebhook:\n  port: 9443\n"}
```

alculquicondor replied:
we could de-duplicate this code by putting it outside of the conditional.

A message such as `Successfully loaded configuration` would cover both codepaths.

tenzen-y replied:
Thanks for the suggestion!
Done.

alculquicondor reviewed (commented):
Final nit

alculquicondor commented on main.go:
```diff
@@ -219,41 +219,37 @@ func apply(configFile string) (ctrl.Options, config.Configuration) {
-	config := config.Configuration{}
-
+	successLoadConfigurationMessage := "Successfully loaded configuration"
 	if configFile == "" {
```
I meant something more like this (take it as pseudocode)

```
if configFile == "" {
  scheme.Default(cfg)
  options = options.AndFrom(cfg)
} else {
  options = options.AndFrom(ctrl.ConfigFile().AtPath(configFile).OfKind(cfg))
}
cfgStr = encodeConfig(cfg)
setupLog("Successfully loaded configuration", "config", cfgStr))
```

tenzen-y replied:
Thanks for clarifying!

tenzen-y replied:
Done.

alculquicondor reviewed (commented):
Thanks!

/lgtm
/approve
